rfcs/2020-05-25-more-usable-logevents.md

```diff
@@ -0,0 +1,223 @@
+
+```rust
+//! serde_json::value::Value
+pub enum Value {
+    Null,
+    Bool(bool),
+    Number(Number),
+    String(String),
+    Array(Vec<Value>),
+    Object(Map<String, Value>),
+}
+```
+
+Differences are:
+
+- `Event` uses a `BTreeMap` , where `serde_json`'s `Map` type is only `BTreeMap` when the `preserve_order` feature flag is turned off. We do not enable this, feature, so they are equivalent.
+- A `Bytes` type in our `Value` type exists because `serde_json::value::Value` and most JSON implementations support this type through the `String` type, which isn't ideal for applying `Bytes` related functions on. We hold this special `Bytes` variant to support better in-pipeline processing for Vector.
+- Our `Value` type lacks a `String` variant, using the `Bytes` variant instead, preventing us from having optimized `String` operations.
+- A `Timestamp` type in our `Value` type exists for largely the same reason the `Bytes` variant exists: Optimize in-pipeline processing.
+- A `Float` and `Integer` type exist where `serde_json::value::Value::Number` exists also exist for optimizing in-pipeline processing.
+
+### `LogEvent` API Functionality of Note
+
+The `LogEvent` API implements a pseudo- `Flatmap` API. It uses special `PathIter` types to parse `String` or `Atom` keys provided to the APIs (`get`, `insert`, etc).
+
+The `Atom` type is deprecated, please use `String` and adapt any `Atom` usages to `String` where possible. See [#1891](https://github.com/timberio/vector/pull/1891).
+
+The current `LogEvent` API is as follows *(documentation, if it exists, is included)*:
+
+```rust
+//! Summarized from src/event/mod.rs
+impl LogEvent {
+    pub fn new() -> Self;
+    pub fn get(&self, key: &Atom) -> Option<&Value>;
+    pub fn get_mut(&mut self, key: &Atom) -> Option<&mut Value>;
+    pub fn contains(&self, key: &Atom) -> bool;
+    pub fn insert<K, V>(&mut self, key: K, value: V) -> Option<Value>
+	    where K: AsRef<str>, V: Into<Value>;
+    pub fn insert_path<V>(&mut self, key: Vec<PathComponent>, value: V) -> Option<Value>
+	    where V: Into<Value>;
+    pub fn insert_flat<K, V>(&mut self, key: K, value: V)
+	    where K: Into<String>, V: Into<Value>;
+    pub fn try_insert<V>(&mut self, key: &Atom, value: V)
+	    where  V: Into<Value>;
+    pub fn remove(&mut self, key: &Atom) -> Option<Value>;
+    pub fn remove_prune(&mut self, key: &Atom, prune: bool) -> Option<Value>;
+    pub fn keys<'a>(&'a self) -> impl Iterator<Item = String> + 'a;
+    pub fn all_fields<'a>(&'a self) -> impl Iterator<
+				Item = (String, &'a Value)
+		> + Serialize;
+    pub fn is_empty(&self) -> bool;
+}
+```
+
+Many of these APIs offer functionality ***coarsely similar*** to a `FlatMap`. These APIs were added ad-hoc, as needed over time, and were not designed intentionally to all fit together. In addition, some changes were added unobtrusively to aid in the migration path.
```

Comment:
Which migration path are you referencing here?

Reply:
`insert_path` etc. I'll add an example. :)